test: Measure and track code coverage

ChristianZaccaria opened the pull request:
# What does this PR do?
<!-- Provide a short summary of what this PR does and why. Link to relevant issues if applicable. -->
- Added coverage badge to README. - [See my fork](https://github.com/ChristianZaccaria/llama-stack)
- Added a GitHub Actions workflow that runs the tests and updates the coverage badge. - [See run](https://github.com/ChristianZaccaria/llama-stack/actions/runs/16203511391/job/45748113233)
- Documented steps in `testing.md` for running the tests locally, and viewing the `html` report.
- Excluded non-essential files from coverage reporting to provide a more accurate measurement.

Automatically created PR to update coverage badge: https://github.com/ChristianZaccaria/llama-stack/pull/9 

# Note for reviewers
1. Currently the coverage report shows a 45% coverage. Wondering if there are other files or directories that should also be excluded from the report to increase the percentage. The directories with the least test coverage are `llama_stack/cli`, `llama_stack/models`, and `llama_stack/ui`. - Should we exclude these?
2. **[Required]** The `GITHUB_TOKEN` should have write permissions to open a PR to update the coverage badge.

# GitHub Issue
<!-- If resolving an issue, uncomment and update the line below -->
<!-- Closes #[issue-number] -->
Closes #2355 

## Test Plan
<!-- Describe the tests you ran to verify your changes with result summaries. *Provide clear instructions so the plan can be easily re-executed.* -->
The `testing.md` file describes how to run the unit tests locally.

leseb commented on .github/workflows/codecov-badge.yaml:
```diff
@@ -0,0 +1,43 @@
+
+      - name: Generate XML coverage report
+        run: |
+          uv run --python ${{ matrix.python }} coverage xml -o htmlcov-${{ matrix.python }}/coverage.xml
```
can we do this in the current unit test job?

leseb replied:
Missed that part thanks

leseb commented on .github/workflows/unit-tests.yml:
```diff
@@ -48,3 +48,16 @@ jobs:
+          coverage=$(uv run --python ${{ matrix.python }} --group unit coverage report -m | tail -1 | tail -c 4 | head -c 2)
+          echo "Total coverage: ${coverage}%"
+
+          # Check if coverage is below the threshold
+          if (( ${coverage} < 40 )); then
+              echo "Coverage failed at ${coverage}%, expected at least 40%"
+              exit 1
+          else
+              echo "Coverage check passed with ${coverage}%"
+          fi
```
Why do we do this?

ChristianZaccaria replied:
It's a tradeoff in exchange for more test coverage. We can remove the check, and maybe we could revisit this in the future.

leseb commented on docs/source/contributing/testing.md:
````diff
@@ -4,3 +4,49 @@ Tests are of three different kinds:
+The easiest way to run unit tests is using the provided script:
+
+```bash
+# Activate virtual environment
+uv venv .venv --python 3.12
+source .venv/bin/activate
+
+# Install dependencies
+uv pip install -r requirements.txt
+
````
all of this is not needed if you run the script

leseb reviewed (changes requested):
Thanks!

ashwinb commented:
Unpopular opinion: I am really not a big fan of code coverage. It leads to low quality unit tests all in the name of increasing a metric. I'd rather have people think through functionality semantically and say "this bit or that interaction" is not covered. Even ask an AI for that. But code coverage as implemented in industry today is detrimental.

ChristianZaccaria commented on .github/workflows/codecov-badge.yaml:
```diff
@@ -0,0 +1,43 @@
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@b9fd7d16f6d7d1b5d2bec1a2887e65ceed900238 # v4.6.0
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Is there someone I could reach out to?

ChristianZaccaria replied:
Thanks @leseb, I've updated the PR. I think the only outstanding item is the `CODECOV_TOKEN` secret.

leseb replied:
ping me offline about this and we can figure it out

ChristianZaccaria commented:
@ashwinb I think that's a fair point, and there's truth to that. It will ultimately be up to reviewers to ensure tests written are up to standards. I think by including code coverage we at least ensure that some thought is given to unit testing.

We can exclude files that don't need to be tested and this way avoid cases where we are writing tests just to tick a box. We can also use `# pragma: no cover` to individual functions to exclude from the report. This way, we focus on unit testing what is important.

leseb commented:
@ChristianZaccaria @ashwinb I understand both perspectives - you both raise good points. I’m fine either way on this.

ChristianZaccaria commented on .github/workflows/coverage-badge.yaml:
```diff
@@ -0,0 +1,62 @@
+      fail-fast: false
+      matrix:
+        python:
+          - "3.12"
```
For this workflow we don't need it. Good call, thanks!

ChristianZaccaria commented on .github/workflows/coverage-badge.yaml:
```diff
@@ -0,0 +1,62 @@
+          git add coverage.svg
+          git commit -m "Updated coverage.svg"
+
+      - name: Create Pull Request
+        if: steps.verify-changed-files.outputs.files_changed == 'true'
+        uses: peter-evans/create-pull-request@271a8d0340265f705b14b6d32b9829c1cb33d45e # v7.0.8
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          title: "ci: [Automatic] Coverage Badge Update"
+          body: |
+            This PR updates the coverage badge based on the latest coverage report.
+
+            Automatically generated by the [workflow coverage-badge.yaml](.github/workflows/coverage-badge.yaml)
+          delete-branch: true
```
This can be checked directly in the repository by an admin.

> 1. Click **Settings** tab 
> 2. On left sidebar click on **Actions** 
> 3. Click **General** 
> 4. scroll down to **Workflow permissions** section 
> 5. select **Read and write permissions**

This setting grants write permissions to workflows running on code that has already been merged into the repository. Meaning, workflows triggered by pull requests do not receive write permissions.

This feature will also be useful for other tasks such as the following to update the list of models automatically via a pull request: https://github.com/meta-llama/llama-stack/issues/2504

ChristianZaccaria commented on .github/workflows/coverage-badge.yaml:
```diff
@@ -0,0 +1,59 @@
+      - 'pyproject.toml'
+      - 'requirements.txt'
+      - '.github/workflows/unit-tests.yml'
+      - '.github/workflows/codecov-badge.yaml' # This workflow
```
Thanks!

ChristianZaccaria commented:
Rebased and updated the documentation.

New successful workflow run: https://github.com/ChristianZaccaria/llama-stack/actions/runs/16290334208/job/45998594216
